qadence/mitigations/readout.py

```diff
@@ -25,51 +51,66 @@ def renormalize_counts(corrected_counts: npt.NDArray, n_shots: int) -> npt.NDArr
         corrected_counts -= counts_diff
         corrected_counts = np.where(corrected_counts < 0, 0, corrected_counts)
         sum_corrected_counts = sum(corrected_counts)
-        if sum_corrected_counts < n_shots:
-            renormalization_factor = max(sum_corrected_counts, n_shots) / min(
-                sum_corrected_counts, n_shots
-            )
-        else:
-            renormalization_factor = min(sum_corrected_counts, n_shots) / max(
-                sum_corrected_counts, n_shots
-            )
+
+        renormalization_factor = n_shots / sum_corrected_counts
         corrected_counts = np.rint(corrected_counts * renormalization_factor).astype(int)
     return corrected_counts
 
 
 def mitigation_minimization(
-    noise: Noise, mitigation: Mitigations, samples: list[Counter]
+    noise: Noise, mitigation: Mitigations, samples: list[Counter], optimization_type: str
 ) -> list[Counter]:
     """Minimize a correction matrix subjected to stochasticity constraints.
 
     See Equation (5) in https://arxiv.org/pdf/2001.09980.pdf.
+    See Page(3) in https://arxiv.org/pdf/1106.5458.pdf for MLE implementation
     """
     noise_matrices = noise.options.get("noise_matrix", noise.options["confusion_matrices"])
     n_qubits = len(list(samples[0].keys())[0])
     n_shots = sum(samples[0].values())
     # Build the whole T matrix.
     T_matrix = reduce(torch.kron, noise_matrices).detach().numpy()
     corrected_counters: list[Counter] = []
+
+    if optimization_type == "mle":  ## MLE version
+        ##check if matrix is singular and use appropriate inverse
+        if np.linalg.matrix_rank(T_matrix) == T_matrix.shape[0]:
```

Comment:
Soft suggestion: make this one a one liner

T_inv = np.linalg.inv(T_matrix) if np.linalg.matrix_rank(T_matrix) == T_matrix.shape[0]  else np.linalg.pinv(T_matrix)